### PR TITLE
Project layout

### LINKED ISSUE #104: Lab 8 - Project Layout
- [ ] Copy the CRUD puppy from upstream master 07_errors/USERNAME
- [ ] Create directory 08_project/USERNAME containing
```
├── README.md
├── go.mod
├── go.sum
├── pkg
│   └── USERNAME-puppy
│       ├── types.go
│       ├── types_test.go
│       ├── errors.go
│       ├── errors_test.go
│       └── store
│           ├── storer.go
│           └── .... store files and tests, e.g. mapstore.go
├── cmd
│   └── USERNAME-puppy-server
│       └── main.go
└── vendor
```
- [ ] Add project introduction and how to build, run & test it to README.md

### PR DESCRIPTION
**Pull request by runnerdave**
Fixes #104

Review of colleague's PR #551

#### Changes proposed in this PR:

 - Copy contents of Lab07
 - Rearrange into packages
 - Write README.md

### CODE REVIEW

File: 08_project/runnerdave/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Puppy server
+
+Storage of Puppies with two different storage methods.
+
+## Pre-requisites
+
+You need to have [Go](https://golang.org/) installed
```
Comment (juliaogris) [inline 08_project/runnerdave/README.md]: link to installation page?

File: 08_project/runnerdave/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+## Build and install
+
+```
+go install ./...
+```
+
+### Run installed version
+
+```
+$GOPATH/bin/puppy-server
```
Comment (juliaogris) [inline 08_project/runnerdave/README.md]: just `puppy-server` with a note, assuming `$GOPATH/bin` is on your path?

**Review by juliaogris — DISMISSED** [review]
LGTM

(just a few nitpicky comments on readme, feel free to ignore).

File: 08_project/runnerdave/README.md (hunk 1 above)
Comment (runnerdave) [reply 08_project/runnerdave/README.md]: fixed

File: 08_project/runnerdave/README.md (hunk 2 above)
Comment (runnerdave) [reply 08_project/runnerdave/README.md]: fixed

File: 08_project/runnerdave/pkg/puppy/store/sync_store.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	return puppy.Puppy{}, puppy.Errorf(puppy.ErrIDNotFound, "puppy with ID:%d not found", id)
+}
+
+// UpdatePuppy puts new puppy data to the store, error if id does not exist
+func (s *SyncStore) UpdatePuppy(id uint16, p *puppy.Puppy) error {
+	if err := puppy.ValidateValue(p.Value); err != nil {
+		return err
+	}
+	if _, ok := s.Load(id); !ok {
```
Comment (juliaogris) [inline 08_project/runnerdave/pkg/puppy/store/sync_store.go:50]: sorry,  I missed this before -   think you should add the mutex here too.

**Review by juliaogris — CHANGES_REQUESTED** [review]
one last little thing I missed before.

File: 08_project/runnerdave/pkg/puppy/store/sync_store.go (hunk 3 above)
Comment (runnerdave) [reply 08_project/runnerdave/pkg/puppy/store/sync_store.go:50]: Sorry @juliaogris you mean to add a lock? I don't understand

Comment (juliaogris) [reply 08_project/runnerdave/pkg/puppy/store/sync_store.go:50]: yes, apologies for the confusion you load and you store - this should happen as an atomic action, so please add
```go
         s.Lock()
	defer s.Unlock()
```
as below.

Comment (runnerdave) [reply 08_project/runnerdave/pkg/puppy/store/sync_store.go:50]: I see, thanks for pointing that out, implemented

**Review by juliaogris — APPROVED** [review]
LGTM - looks really solid imo.

**Pull request merged**